### PR TITLE
Add GuzzleHttp

### PR DESCRIPTION
**Pull request by TimWolla**
I've carefully made sure to reimplement all the various details of HTTPRequest. Specifically I tested all of the following:

- Retrieving a maximum length of the body works properly.
- The response header arrays are identical.
- Any Exception thrown is wrapped in an appropriate legacy Exception with a proper message and code.
- Authenticated requests send the same credentials.
- Redirects are being limited with a proper Exception.
- Requests with a body look identical.

Additionally I made sure that I was able to install a package, including dependencies using the package servers.

All in all I don't expect any surprises and if there's anything then the flexibility of Guzzle should allow to easily fix those.

### CODE REVIEW

**Comment by Hanashi** [comment]
I believe there could be problems after this change if a developer calls `getReply` twice before the change.

**Comment by TimWolla** [comment]
> I believe there could be problems after this change if a developer calls `getReply` twice before this change.

While I did not actually test this I am fairly sure that the implementation is correct, because I explicitly read the body into a string. I'll make sure to test this before merging, though.

**Comment by TimWolla** [comment]
> I'll make sure to test this before merging, though.

I just have verified that calling `getReply()` multiple times will always return identical information.

File: wcfsetup/install/files/lib/system/io/HttpFactory.class.php [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	 * 
+	 * @return ClientInterface
+	 * @see Client
+	 */
+	public static function makeClient(array $options = []) {
+		return new Client(array_merge([
+			'proxy' => PROXY_SERVER_HTTP,
+			'headers' => [
+				'user-agent' => self::getDefaultUserAgent(),
+			]
```
Comment (Gravatronics) [inline wcfsetup/install/files/lib/system/io/HttpFactory.class.php:63]:
```suggestion
			],
```

File: wcfsetup/install/files/lib/util/HTTPRequest.class.php [hunk 2]
```diff
@@ -206,385 +152,102 @@ public function __construct($url, array $options = [], $postParameters = [], arr
...
+			'allow_redirects' => [
+				'max' => $this->options['maxDepth'],
+				'track_redirects' => true,
+				'on_redirect' => $redirectHandler,
+			],
+		];
+		if (isset($this->options['auth'])) {
+			$options['auth'] = [
+				$this->options['auth']['username'],
+				$this->options['auth']['password']
```
Comment (Gravatronics) [inline wcfsetup/install/files/lib/util/HTTPRequest.class.php:179]:
```suggestion
				$this->options['auth']['password'],
```

File: wcfsetup/install/files/lib/util/HTTPRequest.class.php [hunk 3]
```diff
@@ -595,10 +258,40 @@ private function parseReply() {
...
+		
 		return [
-			'statusCode' => $this->statusCode, 
-			'headers' => $this->legacyHeaders,
-			'httpHeaders' => $this->replyHeaders,
+			'statusCode' => (string) $this->response->getStatusCode(), 
+			'headers' => $legacyHeaders,
+			'httpHeaders' => $headers,
 			'body' => $this->replyBody,
 			'url' => $this->url
```
Comment (Gravatronics) [inline wcfsetup/install/files/lib/util/HTTPRequest.class.php:296]:
```suggestion
			'url' => $this->url,
```

**Pull request merged**